gnovm/pkg/gnolang/realm.go

```diff
@@ -453,11 +458,20 @@ func (rlm *Realm) incRefCreatedDescendants(store Store, oo Object) {
 				// NOTE: may already be marked for first gen
 				// newCreated or updated.
 				child.SetOwner(oo)
-				rlm.incRefCreatedDescendants(store, child)
+
+				// There’s a cycle when incrementing the reference count for:
+				// package block -> file block -> package block.
+				// To avoid redundant marking as dirty, set the block to "new real" first
+				// before recursing. This helps identify if it’s already marked as "new real"
+				// in the second visit, eliminating the need for another dirty set.
+				// XXX, side effects?
 				child.SetIsNewReal(true)
+				rlm.incRefCreatedDescendants(store, child)
 			}
 		} else if rc > 1 {
-			rlm.MarkDirty(child)
+			if !isUnsaved(child) {
+				rlm.MarkDirty(child)
+			}
```

Comment:
solve duplicate object issue.

Reply:
another way should be more natural, that mark it dirty after if refCount changes, and remove it from .created to make sure they don't exist in .created .updated in the same time.